<?xml version="1.0" ?>
<!-- THIS FILE LAUNCHES THE FRANKA EMIKA PANDA + SOFTHAND: BOTH THE PACKAGES ARE NEEDED -->
<launch>
    <!-- LAUNCH ARGUMENTS -->

    <!-- Robot to be used -->
  	<arg name="robot_name" default="panda_arm"/>

    <!-- Panda params -->
    <arg name="panda_arm_enabled" default="true"/>
    <arg name="robot_ip" default="172.16.0.102"/>
    <arg name="load_gripper" default="false"/>

    <!-- START LAUNCH -->

    <!-- Robot description -->
    <param name="robot_description" command="$(find xacro)/xacro --inorder '$(find franka_description)/robots/$(arg robot_name).urdf.xacro'" unless="$(arg load_gripper)" />

    <!-- Joint state and robot state publishers of the full robot description -->
	<node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher">
		<!-- FOR USING JOINT STATE PUBLISER FOR GETTING FINGER JOINT STATES -->
		<rosparam param="source_list">[/joint_states_desired]</rosparam>
		<!-- FOR USING FINGER FK PACKAGE FOR GETTING FINGER JOINT STATES -->
		<!-- <rosparam param="source_list">[/panda_arm/joint_states,/all_hand_joint_states]</rosparam> -->
        <param name="rate" value="30"/>
	</node>

    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    
    <!-- Panda namespace -->
    <group ns="panda_arm">

        <!-- A copy of robot description -->
        <param name="robot_description" command="$(find xacro)/xacro --inorder '$(find franka_description)/robots/$(arg robot_name).urdf.xacro'" unless="$(arg load_gripper)" />

        <!-- Panda hardware interface -->
        <group if="$(arg panda_arm_enabled)">
            <node name="franka_control" pkg="franka_control" type="franka_control_node" output="screen" required="true">
                <rosparam command="load" file="$(find franka_control)/config/franka_control_node.yaml" />
                <param name="robot_ip" value="$(arg robot_ip)" />
            </node>
        </group>

        <!-- Panda controllers -->
        <rosparam command="load" file="$(find franka_control)/config/default_controllers.yaml" />
        <rosparam command="load" file="$(find panda_controllers)/config/panda_controllers.yaml" />
        <rosparam command="load" file="$(find franka_softbots)/config/franka_softbots.yaml" />
        <node name="state_controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="franka_state_controller"/>
        <node name="controller_spawner" pkg="controller_manager" type="controller_manager" respawn="false" output="screen"  args="spawn pd_controller"/>
        <node name="controller_stopper" pkg="controller_manager" type="controller_manager" respawn="false" output="screen"  args="load  position_joint_trajectory_controller computed_torque_controller back_stepping_controller"/>
        
        <!-- Panda joint state publisher -->
        <node name="joint_state_publisher" type="joint_state_publisher" pkg="joint_state_publisher" output="screen">
            <rosparam if="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states, franka_gripper/joint_states] </rosparam>
            <rosparam unless="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states] </rosparam>
            <param name="rate" value="30"/>
        </node>
        <!-- Panda desired state publisher -->
        <node name="joint_state_desired_publisher" type="joint_state_publisher" pkg="joint_state_publisher" output="screen">
            <rosparam if="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states_desired, franka_gripper/joint_states] </rosparam>
            <rosparam unless="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states_desired] </rosparam>
            <param name="rate" value="30"/>
            <remap from="/panda_arm/joint_states" to="/joint_states_desired" />
        </node>

    </group>

    <!-- RViz for visualization -->
    <!-- <node pkg="rviz" type="rviz" output="screen" name="rviz" args="-d $(find franka_visualization)/launch/franka_visualization.rviz"/> -->

</launch>
